libgeos/libgeos_handle.go

```diff
@@ -0,0 +1,236 @@
+package libgeos
+
+/*
+#cgo linux CFLAGS: -I/usr/include
+#cgo linux LDFLAGS: -L/usr/lib -lgeos_c
+#include "geos_c.h"
+#include <stdlib.h>
+#include <string.h>
+
+void sf_error_handler(const char *message, void *userdata) {
+	strncpy(userdata, message, 1024);
+}
+
+GEOSContextHandle_t sf_init(void *userdata) {
+	GEOSContextHandle_t ctx = GEOS_init_r();
+	GEOSContext_setErrorMessageHandler_r(ctx, sf_error_handler, userdata);
+	return ctx;
+}
+
+*/
+import "C"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/peterstace/simplefeatures/geom"
+)
+
+// Handle is an opaque handle that can be used to invoke libgeos operations.
+// Instances are not threadsafe and thus should not be used in a concurrent
+// manner.
+type Handle struct {
+	context C.GEOSContextHandle_t
+	errBuf  [1024]byte
+}
+
+// NewHandle creates a new libgeos handle.
+func NewHandle() (*Handle, error) {
+	h := &Handle{}
+	h.context = C.sf_init(unsafe.Pointer(&h.errBuf))
+	if h.context == nil {
+		return nil, errors.New("could not create libgeos context")
+	}
+	return h, nil
+}
+
+// err gets the last error message reported by libgeos as an error type. It
+// always returns a non-nil error. If no error message has been reported, then
+// it returns a generic error message.
+func (h *Handle) err() error {
+	msg := h.errMsg()
+	if msg == "" {
+		// No error stored, which indicates that the error handler didn't get
+		// trigged. The best we can do is give a generic error.
+		msg = "libgeos internal error"
+	}
+	h.errBuf = [1024]byte{} // Reset the buffer for the next error message.
+	return errors.New(strings.TrimSpace(msg))
+}
+
+// errMsg gets the textual representation of the last error message reported by
+// libgeos.
+func (h *Handle) errMsg() string {
+	// The error message is either NULL terminated, or fills the entire buffer.
+	firstZero := len(h.errBuf)
+	for i, b := range h.errBuf {
+		if b == 0 {
+			firstZero = i
```

Comment:
Or you could omit the `firstZero` variable entirely and return `string(h.errBuf[:i])` here and return `string(h.errBuf)` later.

Reply:
Good suggestion, that's a bit easier to understand. Fixed.